More minor code for nodegroup elements

nilsdeppe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add nodegroup support to InboxTags

nilsdeppe committed:
Add BoundaryData tag

nilsdeppe committed:
Add get_node_lock to distributed object

nilsdeppe committed:
Support nodegroup element array in DistributedObject

nilsdeppe committed:
Initialize BoundaryData in mortar initialization

kidder commented on src/Evolution/DiscontinuousGalerkin/InboxTags.hpp:
```diff
@@ -236,6 +237,19 @@ struct BoundaryCorrectionAndGhostCellsInbox {
+ *
+ * Since the inbox data can be received asynchronously and lockfree ordered
+ * containers are slow and challenging to implement, we instead using an
+ * unordered container for the inbox, then transfer the data into an order map
```
ordered

knelli2 replied:
[nitpick] also `using` -> `use`

kidder reviewed (changes requested):
you can squash

knelli2 commented on src/Evolution/DiscontinuousGalerkin/InboxTags.hpp:
```diff
@@ -109,19 +113,47 @@ struct BoundaryCorrectionAndGhostCellsInbox {
+  static size_t insert_into_inbox(const gsl::not_null<type_spsc*> inbox,
+                                  const temporal_id& time_step_id,
+                                  ReceiveDataType&& data) {
+    const DirectionalId<Dim>& neighbor_id = data.first;
+    // Note: This assumes the neighbor_id is oriented into our (the elment
```
[nitpick] `elment` -> `element`

knelli2 commented on src/Evolution/DiscontinuousGalerkin/InboxTags.hpp:
```diff
@@ -109,19 +113,47 @@ struct BoundaryCorrectionAndGhostCellsInbox {
+  // The actual type being used.
+  using type = type_map;
 
+  template <typename ReceiveDataType>
+  static size_t insert_into_inbox(const gsl::not_null<type_spsc*> inbox,
```
You've changed the return type of this function. Could you add some docs about what the `size_t` represents? From the looks of it, it's just the new size of the inbox

nilsdeppe commented:
I've rebased and pushed fixups. I added a decent number of words for the return type being `size_t` since that is important. I'll need to clean this class up a bit after everything is working, mostly just the documentation. I also added `value_type` as a type alias since this is actually what the STL uses. It's not used right now, though.

kidder reviewed (commented):
fixups LGTM

knelli2 reviewed (approved):
Fixups look good. Squash and rebase

nilsdeppe commented:
Done! Thanks!
